### PR TITLE
Store more source location information in Boogie AST

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
This is a refactoring that enables other features

### How has this been tested?
This refactoring is covered by existing tests

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Verifier/Translator.cs [hunk 1]
```diff
@@ -3471,8 +3471,15 @@ public enum StmtType { NONE, ASSERT, ASSUME };
...
-        return node.RangeToken.ToToken();
+        return new BoogieRangeToken(node.StartToken, node.EndToken, node.Tok);
       } else {
+        // The commented line is what we want, but it changes what is translated.
+        // Seems to relate to refinement and possibly RefinementToken.IsInherited and or ForceCheckToken
+        // It might be better to remove calls to RefinementToken.IsInherited from this file, and instead
+        // add generic attributes like {:verify false} in the refinement phases, so that refinement does not complicate
+        // translation,
+        //
+        // return new BoogieRangeToken(node.StartToken, node.EndToken, node.Tok);
```
Comment (MikaelMayer) [inline Source/DafnyCore/Verifier/Translator.cs:3482]: That's precisely what I encountered. I think there might be two mechanism for reporting related ranges: One in nested token, another one in auxiliary information retrieved from Boogie. That might create a conflict that we solve one way or the other, but it's incompatible with the correct code above.
At least, given that we have tests in place, it should be enough for us to still have a decent behavior.

**Review by MikaelMayer — APPROVED** [review]
That's great and very similar to what I suggested, so I approve it without hesitation. I will also reapprove as soon as you find how to fix the CI test. Thank you very much for the effort!

File: Source/DafnyCore/Rewriters/RefinementTransformer.cs [hunk 2]
```diff
@@ -58,7 +58,8 @@ public RefinementToken(IToken tok, ModuleDefinition m)
       }
 
       var rtok = tok as RefinementToken;
-      return rtok != null && rtok.InheritingModule == m;
+      var result = rtok != null && rtok.InheritingModule == m;
+      return result;
```
Comment (MikaelMayer) [inline Source/DafnyCore/Rewriters/RefinementTransformer.cs:62]: Non-blocking: this refactoring introduces a variable that does not seem to provide meaning to the result

**Review by MikaelMayer — APPROVED** [review]
Good job for fixing the tests, it wasn't easy

**Review by MikaelMayer — DISMISSED** [review]

**Review by stefan-aws — DISMISSED** [review]

**Pull request merged**